simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java

```diff
@@ -11,44 +14,60 @@
 class TimeWindowQuantiles {
 
   private final Quantile[] quantiles;
-  private final CKMSQuantiles[] ringBuffer;
-  private int currentBucket;
-  private long lastRotateTimestampMillis;
+  private final ConcurrentLinkedQueue<CKMSQuantiles> buckets;
+  private final AtomicLong lastRotateTimestampMillis;
   private final long durationBetweenRotatesMillis;
 
   public TimeWindowQuantiles(Quantile[] quantiles, long maxAgeSeconds, int ageBuckets) {
     this.quantiles = quantiles;
-    this.ringBuffer = new CKMSQuantiles[ageBuckets];
+    this.buckets = new ConcurrentLinkedQueue<CKMSQuantiles>();
     for (int i = 0; i < ageBuckets; i++) {
-      this.ringBuffer[i] = new CKMSQuantiles(quantiles);
+      this.buckets.add(new CKMSQuantiles(quantiles));
     }
-    this.currentBucket = 0;
-    this.lastRotateTimestampMillis = System.currentTimeMillis();
+    this.lastRotateTimestampMillis = new AtomicLong(System.currentTimeMillis());
     this.durationBetweenRotatesMillis = TimeUnit.SECONDS.toMillis(maxAgeSeconds) / ageBuckets;
   }
 
   public double get(double q) {
-    CKMSQuantiles currentBucket = rotate();
+    // On concurrent `get` and `rotate`:
+    //  - it is acceptable to `get` the sample from an outdated `bucket`.
+    //  - `currentBucket` could be `null` when there is only a single bucket (edge case).
+    rotate();
+
+    CKMSQuantiles currentBucket;
+    do {
+      currentBucket = buckets.peek();
+    } while (currentBucket == null);
```

Comment:
It should be fine to suspend the thread during busy waiting with `Thread.yield()`.